File: server/player_selection_test.go
package main

import "testing"

func TestPossessionSelectionReference(t *testing.T) {
	for _, period := range []int{1, 2} {
		for team := 0; team < 2; team++ {
			for _, facing := range [][2]float64{{1, 0}, {-1, 0}, {0, 1}, {0, -1}} {
				s := initial()
				s.Period = period
				for i := range s.Players {
					s.Players[i].X, s.Players[i].Z = 20, 10
				}
				owner, a := team*9+7, (1-team)*9+6
				b := a + 1
				p := &s.Players[owner]
				p.X, p.Z, p.FX, p.FZ = 0, 0, facing[0], facing[1]
				advancePhysicalPose(p, owner, period, .04)
				ox, oz := physicalBallOffset(p, &s.Ball)
				s.Ball.Owner, s.Ball.X, s.Ball.Z = owner, ox, oz
				s.Players[a].X, s.Players[a].Z = 0, 0
				s.Players[b].X, s.Players[b].Z = ox, oz
				distances := s.possessionDistances()
				if distances[a] != 0 || distances[b] <= 0 {
					t.Fatal("carrier target", distances[a], distances[b])
				}
				s.selectPlayers()
				if s.Controlled[1-team] != a || s.Controlled[team] != owner {
					t.Fatal("held selection", s.Controlled)
				}
				s.Ball.Owner = -1
				s.selectPlayers()
				if s.Controlled[1-team] != b {
					t.Fatal("loose selection", s.Controlled)
				}
			}
		}
	}
}
func TestFallenNearestSelection(t *testing.T) {
	for team := 0; team < 2; team++ {
		s := initial()
		for i := range s.Players {
			s.Players[i].X, s.Players[i].Z = 20, 10
		}
		a := team*9 + 7
		b := a + 1
		s.Ball.Owner, s.Ball.X, s.Ball.Z = -1, 0, 0
		s.Players[a].X, s.Players[a].Z, s.Players[a].Stun, s.Players[a].Action = 0, 0, .5, 4
		s.Players[b].X, s.Players[b].Z = 2, 0
		s.selectPlayers()
		if s.Controlled[team] != a {
			t.Fatal("fallen selection", s.Controlled)
		}
		s.Players[b].X = 0
		s.selectPlayers()
		if s.Controlled[team] != b {
			t.Fatal("tie selection", s.Controlled)
		}
	}
}

File: src/player-selection.test.js
import test from 'node:test';
import assert from 'node:assert/strict';
import { initial, selectPlayers, possessionDistances } from './game.js';
import { advancePhysicalPose, physicalBallOffset } from './physical-pose.js';

function fixture(period) {
  const s = initial();
  s.period = period;
  for (const p of s.players) Object.assign(p, {x: 20, z: 10, stun: 0});
  return s;
}
test('held-ball selection targets carrier terrain across teams, halves and facing', () => {
  for (const period of [1, 2]) for (const team of [0, 1]) for (const [fx,fz] of [[1,0],[-1,0],[0,1],[0,-1]]) {
    const s = fixture(period), owner = team * 9 + 7, a = (1-team)*9+6, b = a+1;
    Object.assign(s.players[owner], {x:0,z:0,fx,fz});
    advancePhysicalPose(s.players[owner], owner, period, .04);
    const offset = physicalBallOffset(s.players[owner], s.ball);
    Object.assign(s.ball, {owner, x:offset.x, z:offset.z});
    Object.assign(s.players[a], {x:0,z:0});
    Object.assign(s.players[b], {x:offset.x,z:offset.z});
    const distances = possessionDistances(s);
    assert.equal(distances[a], 0);
    assert.ok(distances[b] > 0);
    selectPlayers(s);
    assert.equal(s.controlled[1-team], a);
    assert.equal(s.controlled[team], owner);
    s.ball.owner = -1;
    selectPlayers(s);
    assert.equal(s.controlled[1-team], b, 'loose ball becomes the distance target');
  }
});
test('nearest fallen player remains selected and equal distances favor later roster slot', () => {
  for (const team of [0,1]) {
    const s = fixture(1), a = team*9+7, b = a+1;
    Object.assign(s.ball, {owner:-1,x:0,z:0});
    Object.assign(s.players[a], {x:0,z:0,stun:.5,action:4});
    Object.assign(s.players[b], {x:2,z:0});
    selectPlayers(s);
    assert.equal(s.controlled[team], a);
    s.players[b].x = 0;
    selectPlayers(s);
    assert.equal(s.controlled[team], b);
  }
});
